### PR TITLE
fix: Queries with no data points

### LINKED ISSUE #41: Handle queries for which no data points are available
From https://github.com/keptn/keptn/issues/1496

Currently, if we receive a result from Prometheus containing an empty array, such as the following:

```
{
    "status": "success",
    "data": {
        "resultType": "vector",
        "result": []
    }
}
```

we assume that the value for the queried metric is 0.

**Example 1:** If the error rate gets queried and the error rate of the service was 0, the Prometheus API will send the object listed above and prometheus-service will interpret it as a `0`.

**Example 2:** If the response time gets queried, but Prometheus sends the empty result from above, prometheus-service interprets it as a `0`.

Based on the query, the result has different semantics; i.e. it could mean that no data is available and we cannot retrieve a value.

A solution is proposed in https://github.com/keptn-contrib/prometheus-service/issues/41#issuecomment-1109382656

### PR DESCRIPTION
**Pull request by Raffy23**
## This PR
- fixes #41
- marks SLI value queries with zero, more than one or invalid values as failed (Success: false) 
<br/>

With this fix the result field of the emitted `.finished` event changes accordingly: 
- if at least one SLI query fails the cloud event result will be a warning
- if all SLI queries fail the result will be failure

### CODE REVIEW

File: eventhandling/getSliEvent.go [hunk 1]
```diff
@@ -47,33 +47,82 @@ func (eh GetSliEventHandler) HandleEvent() error {
...
+	if len(sliResults) > 0 {
+		sliResultsFailed := 0
+		for _, sliResult := range sliResults {
+			if !sliResult.Success {
+				sliResultsFailed++
+			}
+		}
+
+		if sliResultsFailed == 1 && len(sliResults) > 1 {
+			finalSLIEventResult = keptnv2.ResultWarning
```
Comment (christian-kreuzberger-dtx) [inline eventhandling/getSliEvent.go:115]: Not sure if this is what we want/need.
I think it should be `sliResultsFailed > 0 && sliResultsFailed < len(sliResults)`

File: eventhandling/getSliEvent.go [hunk 2]
```diff
@@ -148,13 +161,6 @@ func retrieveMetrics(eventData *keptnv2.GetSLITriggeredEventData, keptnHandler *
 				Success: false,
 				Message: err.Error(),
 			})
-		} else if math.IsNaN(sliValue) {
-			sliResults = append(sliResults, &keptnv2.SLIResult{
-				Metric:  indicator,
-				Value:   0,
-				Success: false,
-				Message: "SLI value is NaN",
-			})
```
Comment (christian-kreuzberger-dtx) [inline eventhandling/getSliEvent.go:157]: Do we no longer need this case?

Comment (christian-kreuzberger-dtx) [reply eventhandling/getSliEvent.go:157]: Never mind, I see this is now handled in `GetSLIValue` with a `|| math.IsNaN(floatValue)`

**Review by christian-kreuzberger-dtx — APPROVED** [review]
LGTM

**Review by christian-kreuzberger-dtx — APPROVED** [review]
LGTM

File: eventhandling/getSliEvent.go [hunk 3]
```diff
@@ -47,33 +47,82 @@ func (eh GetSliEventHandler) HandleEvent() error {
 
 	// send started event
 	_, err = eh.keptnHandler.SendTaskStartedEvent(eventData, utils.ServiceName)
-
 	if err != nil {
-		errMsg := fmt.Sprintf("Failed to send task started CloudEvent (%s), aborting...", err.Error())
-		log.Println(errMsg)
+		errMsg := fmt.Errorf("failed to send task started CloudEvent: %w", err)
+		log.Println(errMsg.Error())
 		return err
```
Comment (pchila) [inline eventhandling/getSliEvent.go:53]: Please return the wrapping error...
furthermore if an error is returned, no need to log it, the code that will handle the error will decide if logging it or doing something else with it

```suggestion
		return fmt.Errorf("failed to send task started CloudEvent: %w", err)
```

**Pull request merged**